### DISCUSSION TITLE
Tailwind v4 not working with Safari 15

### DISCUSSION THREAD

**Discussion by nevil2** [tailwindlabs/tailwindcss#15284] [Help]
I've updated a project to v4.
It is working on most browsers but does not work on Safari 15.6.1 running on Catalina 10.15.7.
Should/can it work, or is Tailwind dropping support for older browsers.

**Comment by adamwathan** [comment]
Hey! Yeah v4 targets Safari 16.4 and up, has some important features we depend on like `@property` and `color-mix(…)` and since we only do a major version every 2-3 years we decided to target pretty modern browsers to keep the shelf life of v4 as high as possible. If you need to support older browsers I'd recommend using v3 until you can drop support 👍

**Reply by ansf** [reply]
@adamwathan are there any plans on how long v3 will be supported (at least for fixing security issues and compatibility with the ecosystem)?

**Reply by Htbaa** [reply]
Yes, would like to know that as well. Is there an EOL date in mind for v3, or has it already been reached?

**Reply by Neugierdsnase** [reply]
Thanks for the clarification. Are `@property` and `color-mix()` the only features that aren't supported by older Safari versions? Because it might be fairly easy to write a polyfill for the former and a fallback for the latter?

**Reply by fabb** [reply]
A fallback for `@property` would be great. I can live without `color-mix` and/or write fallbacks by hand.

**Reply by alexchexes** [reply]
Wait, does that mean that sites built with Tailwind v4 just won't work (not guarantee to work) on devices from before 2023?

**Reply by frederikhors** [reply]
> Wait, does that mean that sites built with Tailwind v4 just won't work (not guarantee to work) on devices from before 2023?

This is exactly what we're discussing about: https://github.com/tailwindlabs/tailwindcss/discussions/15284#discussioncomment-12248311.

**Reply by alexchexes** [reply]
> This is exactly what we're discussing about

Yeah, I see that, but I just can't figure out which generations of devices we're talking about. Here, I see @adamwathan mentioning that the minimum required version is Safari 16.4, which was released in 2023. That seems way too recent—I can't believe that anyone who doesn't want to bury their library would just cut off most users worldwide.

So, am I understanding this correctly, or am I missing something?

**Reply by frederikhors** [reply]
> am I understanding this correctly

You are. We need a fix for this. Tailwind is awesome but we need to support old browsers too.

**Reply by kieranstartup** [reply]
Just ran into this testing a site on Safari 15.6 this morning and wondering why half the site is broken. Going to downgrade to v3 for the time being.

**Reply by adamwathan** [reply]
> Wait, does that mean that sites built with Tailwind v4 just won't work (not guarantee to work) on devices from before 2023?

iOS 16.4 is supported as far back as the iPhone 8 which was released in 2017.

> are there any plans on how long v3 will be supported (at least for fixing security issues and compatibility with the ecosystem)?

We'll definitely release patch releases as necessary for major issues for at least the next 18 — 24 months, for example making sure it continues to work with popular frameworks like Next.js or new versions of PostCSS, webpack, etc. We won't be adding any new features or fixing long-standing minor bugs that have workarounds or anything like that though, just compatibility issues.

**Reply by alexchexes** [reply]
> iOS 16.4 is supported as far back as the iPhone 8 which was released in 2017

While the share of iOS users on pre‑16.4 versions is globally less than 0.5%, that percentage still represents hundreds of thousands (if not millions) of people, given the sheer scale of internet users worldwide.

Moreover, in poorer countries, the dynamics is quite different - users often hold onto their devices for longer periods, so the proportion of Safari users on older versions could be noticeably higher.

It’s totally reasonable to drop support for advanced features on older browsers, as long as there’s a fallback that prevents the site from becoming completely unusable. For example, it might be acceptable if certain styles degrade - like colors appearing off or advanced interactive features not working. But by removing colors (which are essential for users to actually see the site) we're essentially rendering all Tailwind v4 websites unusable for hundreds of thousands of users worldwide. That level of degradation is not acceptable on a library level, I.M.O.

**Reply by FrenzyExists** [reply]
I was almost loosing my mind trying to figure out why i could not get tailwind to work on safari until i got here. 

I agree with the above comment, its ridiculous to drop support on safari browsers, for me I'm on a 2019 mac with macos 14.5. I don't know you but 2019, 2020, and even as recent as 2023 isn't really a long time ago. Maybe in the US you get a phone every year but that's not a thing everywhere else in the world. Lots of peeps hold to their devices 7, 8 even 10+ years. Why now make a functional site now useless because not even colors work?

**Reply by ansf** [reply]
@adamwathan 

> We'll definitely release patch releases as necessary for major issues for at least the next 18 — 24 months, for example making sure it continues to work with popular frameworks like Next.js or new versions of PostCSS, webpack, etc. We won't be adding any new features or fixing long-standing minor bugs that have workarounds or anything like that though, just compatibility issues.

That's good news. Thanks 👍

**Reply by sdavids** [reply]
https://github.com/tailwindlabs/tailwindcss.com/issues/2102

It would be nice if this info were not buried in a discussion about Safari 15.

**Comment by Azarattum** [comment]
Safari is such a pain to update. Especially on phones where it is linked to your iOS version which is capped by your phone model. So, to update the browser you **literally have to buy a new phone**!

**Reply by hdy1023** [reply]
Expect v4 to come out with a plugin for compatibility with legacy browsers, such as oklch,color-mix,css-nesting.because there are too many changes from v3 to v4, especially `@theme`, a lot of things need to be done to fall back to v3 at this point.

**Reply by Neugierdsnase** [reply]
What do you mean with "expect v4 to come out"? V4 is out and such a plugin does not exist (yet?).

**Reply by hdy1023** [reply]
I mean can v4 provide a plugin or config to support legacy browsers like v3?

**Reply by frederikhors** [reply]
This is an enormous issue for us. Any plugin there yet?

**Reply by frederikhors** [reply]
What do you think about using https://lightningcss.dev? Can it solve this issue? Maybe using it with ViteJS bundling for Safari older versions only loading correct css at runtime?

**Reply by frederikhors** [reply]
I tried adding lightningcss to a SvelteKit (Vite) project using [the docs setup](https://vite.dev/guide/features#lightning-css) but I get this error:

```
[lightningcss] Invalid token in pseudo element: WhiteSpace(" ")

animation: var(--animate-spin);
```

Why?

**Reply by Azarattum** [reply]
I thought Tailwind v4 already uses Lightning CSS under the hood, doesn't it?

**Reply by Neugierdsnase** [reply]
According to the docs it does, but only to flatten nesting:  https://tailwindcss.com/docs/compatibility#nesting

**Reply by Neugierdsnase** [reply]
It really blocks us from upgrading right now, we are not willing to sacrifice the users currently still on these versions and there is no graceful fallback, it just completely breaks the site.

**Reply by frederikhors** [reply]
> It really blocks us from upgrading right now, we are not willing to sacrifice the users currently still on these versions and there is no graceful fallback, it just completely breaks the site.

Yeah. This is a real issue. For example many iPad or iPhone devices are completely cut off! There is no possibility of updating them and yet they work very well and are still used a lot!

**Reply by alexchexes** [reply]
> many

Maybe some actual numbers would help. I tried to find statistics quickly but haven't had any luck yet. I'm curious about the share of users with unsupported browsers, especially outside the US/Canada.

**Reply by fabb** [reply]
Browserslist says 0.47% world-wide, 0.88% north america, 0.56% europe:
https://browsersl.ist/#q=Safari+%3C+16.4

**Reply by Azarattum** [reply]
There is a difference though in cutting a percentage of users who just need to hit an "update" button and who would literally have no way to access your website anymore.

If an app would do this, I could get their decision. But doing this on a library level is much much worse. You are making this decision for thousands of developers and their apps who may have millions of users. Even small percentages become big numbers here.

**Reply by frederikhors** [reply]
Yeah. These numbers say nothing. We have hundreds of devices in retail stores using not upgradable Safari 15.

**Reply by fabb** [reply]
I agree. There’s a difference between a page with a few visual issues on older devices and a completely unusable page.

**Reply by davecarlson** [reply]
major issues for us have been:
- invisible buttons ( as the bg-color isn't showing, but text-white DOES work)
- invisible inputs ( due to the border- not working, where the input fields are transparent background or white background with borders) .
AS mentioned lower down, `@csstools/postcss-oklab-function` and `postcss-nested`  have improved things somewhat, but it's by no means perfect

**Comment by alexchexes** [comment]
<img src="https://github.com/user-attachments/assets/ff31f07a-96ce-4bf8-a040-d9e06c368046" height="360">
<img src="https://github.com/user-attachments/assets/49effb4f-fa27-49f0-8603-53a3866a8263" height="360">

The second one is Safari 15.6.

This is just insane.

Rolling back to v3, but it's too bad that we had already updated to v4 before realizing that it no longer works for a broader audience. Upgrading is easier than rolling back, there's no "v4-to-v3" migration tool, sleepless nights are ahead...  
Also, some libraries, like `tailwind-merge`, have **dropped support for v3**, making things even worse... How did we get to this point

**Reply by frederikhors** [reply]
> Rolling back to v3, but it's too bad that we had already updated to v4 before realizing that it no longer works for a broader audience. Upgrading is easier than rolling back, there's no "v4-to-v3" migration tool, sleepless nights are ahead...
> Also, some libraries, like `tailwind-merge`, have **dropped support for v3**, making things even worse... How did we get to this point

We're in the exact same situation.

Instead of rolling back, how about we join forces to create a workaround? I tried using [lightningcss](https://github.com/tailwindlabs/tailwindcss/discussions/15284#discussioncomment-12206236) but there are issues. Can you help?

**Reply by alexchexes** [reply]
> lightningcss

It doesn't seem to transpile any of Tailwind’s "cutting-edge" CSS. For instance, all these lines from the Tailwind v4 output remain unchanged:
```
--color-red-500: oklch(.637 .237 25.331);

color: color-mix(in oklab, var(--color-slate-500) 5%, transparent);

color: color-mix(in oklab, currentColor 50%, transparent);
```

[playground](https://lightningcss.dev/playground/index.html#%7B%22minify%22%3Afalse%2C%22customMedia%22%3Afalse%2C%22cssModules%22%3Afalse%2C%22analyzeDependencies%22%3Afalse%2C%22targets%22%3A%7B%22safari%22%3A983040%7D%2C%22include%22%3A1113088%2C%22exclude%22%3A0%2C%22source%22%3A%22%3Aroot%20%7B%5Cn%20%20--color-red-500%3A%20oklch(.637%20.237%2025.331)%3B%5Cn%7D%5Cn%5Cn.foo%20%7B%5Cn%20%20color%3A%20color-mix(in%20oklab%2Cvar(--color-red-500)5%25%2Ctransparent)%5Cn%7D%5Cn%5Cn.bar%20%7B%5Cn%20%20color%3A%20color-mix(in%20oklab%2CcurrentColor%2050%25%2Ctransparent)%5Cn%7D%22%2C%22visitorEnabled%22%3Afalse%2C%22visitor%22%3A%22%7B%5Cn%20%20Color(color)%20%7B%5Cn%20%20%20%20if%20(color.type%20%3D%3D%3D%20'rgb')%20%7B%5Cn%20%20%20%20%20%20color.g%20%3D%200%3B%5Cn%20%20%20%20%20%20return%20color%3B%5Cn%20%20%20%20%7D%5Cn%20%20%7D%5Cn%7D%22%2C%22unusedSymbols%22%3A%5B%5D%2C%22version%22%3A%22local%22%7D)

**Reply by hdy1023** [reply]
> We're in the exact same situation.
> 
> Instead of rolling back, how about we join forces to create a workaround? I tried using [lightningcss](https://github.com/tailwindlabs/tailwindcss/discussions/15284#discussioncomment-12206236) but there are issues. Can you help?

I have resolved the issues related to `oklch`, `oklab`, and `css-nesting` by using the plugins `@csstools/postcss-oklab-function` and `postcss-nested` of `PostCSS` . Currently, `theme color` and `space-x-*` are working on Safari 15.

I'm still trying to find a solution for the `color-mix` problem. Although `PostCSS` has a plugin (@csstools/postcss-color-mix-function), it can only handle static values and is unable to deal with `Css Var`. 

If you have any good ideas, you are welcome to share and communicate them.

**Reply by fabb** [reply]
@hdy1023 awesome, thank you for sharing!
Have you found a way to backfill `@property`?

**Reply by hdy1023** [reply]
> @hdy1023 awesome, thank you for sharing! Have you found a way to backfill `@property`?

I find that `@property` is not used much.
If it has impacts in your project, you can manually replace it with `CssVar` and assign a default value.

**Comment by devhe4d** [comment]
I have css issues with Safari 18 on macOS Sequoia (15.3.1) even. the `bg-xxx-xxx` doesn't work what does that mean? should I migrate everything to v3? couldn't we have some sort of polyfills ?

update #1: for every change to reflect on Safari 18, I need to to CMD+OPTION+R to hard refresh the page ( background appears once I do this )

Update #2: the issue is Turbopack, if you are using Next.js + Turbopack, change it to Webpack instead and it will work just fine

**Comment by shashanksrajak** [comment]
It does not work with Safari v18.1.1 as well, I just setup NextJS with Tailwind v4 and used safari to test out, but it does not work. While it worked on Google Chrome.

**Reply by devhe4d** [reply]
> It does not work with Safari v18.1.1 as well, I just setup NextJS with Tailwind v4 and used safari to test out, but it does not work. While it worked on Google Chrome.

the issue is Turbopack, if you are using Next.js + Turbopack, change it to Webpack instead and it will work just fine

**Reply by zulfiismailovdemiri** [reply]
just remove class name "relative" from body or there where you use bg-gradient-to-* from-*-* to-*-*

**Comment by Eptagone** [comment]
I also had some problems like borders not showed (previously reported in #15506) and colors but fortunely, I fixed them using **lightningcss** and custom transformers. Sadly, there're still some issues with the `custom-mix` function and maybe others that I haven't detected yet.

I've published a [gist here](https://gist.github.com/Eptagone/a18be9adabaf8ecc54d1c4e6337c95b2) in case it helps.

**Comment by Azarattum** [comment]
https://github.com/tailwindlabs/tailwindcss/pull/17435 seems promising. Has anyone from this thread tested it yet? What features are still missing if we upgrade from v3 to v4 for Safari 15? Or is v4 now fully backwards compatible with this PR?

**Comment by mustafa0x** [comment]
The migrate script (`@tailwindcss/upgrade`) really should've had a **HUGE RED WARNING** that this'll break the site/app for anyone with a browser more than 3 years old. :(

And we've only noticed issues a few weeks later, so we can't simply `git reset --hard HEAD^`.

I appreciate that the team is working on compat with the 4.1.0 and later releases.

**Reply by RobinMalfait** [reply]
Update to the latest version and you should be good to go 👍 

https://tailwindcss.com/blog/tailwindcss-v4-1#improved-compatibility-with-older-browsers

**Reply by mustafa0x** [reply]
Thanks @RobinMalfait! Is Chrome 87 supported?

**Reply by RobinMalfait** [reply]
Hmm that's a different story. Obviously a lot of features that aren't available in the browser won't work. Colors degrade way better now, but probably not as far as Chrome 87. That said, I assume if you want to support old browsers like that, that you are already working with hex colors instead of more vibrant colors.

So in short, it's going to depend on what you use right now. So I would suggest to bump to the latest v4.1.* version which should already have much better support, but you will have to test your app/website to see where things are still lacking.

**Comment by m-nathani** [comment]
Apologies for the mini rant, but I wanted to share some frustration we've encountered—this situation really reminded me of the _Internet Explorer_ days... 😿

We recently ran into serious issues with **Safari 13, 14, and 15**. While Safari 16.4 has been out for over two years now, we were surprised to find that our very basic landing page — using simple HTML/CSS, a background image, and a few fixed-position elements — completely broke on older Safari versions.

You can see the widget here:
👉 https://reservation.umai.io/en/widget/ronin-izakaya-by-kabuki

There’s nothing advanced or browser-specific here, so it was quite unexpected. Given how foundational Tailwind CSS is, we assumed it would offer strong backward compatibility, especially since modern development encourages progressive enhancement.

Unfortunately, we only discovered this compatibility issue after shipping the update, when customers started reporting broken views. It was a tough debugging experience, and it took us a while to trace it back to the Tailwind upgrade.

We didn't find any compatibility notices prominently highlighted in the release notes or upgrade guides, which made it even harder to anticipate.

This experience is a reminder that while frameworks bring many advantages, unexpected regressions on legacy browsers can have serious real-world consequences.

It would be great if future release notes could include clearer information about browser support changes — especially for older but still-in-use browsers like Safari 13–15 and **atleast allow supported styling to render on legacy browser, rather complete dropship**

Thanks for all the hard work — Tailwind is fantastic overall, and we appreciate the continued improvements.

**Comment by choonx99** [comment]
Hi, i really appreciate the work done at tailwind. So much so that i migrated my project from using sass modules to tailwind v4. However, after migrating, my web page started breaking for mobile users on the older safari versions. I found out now that tailwind v4 does not support responsiveness for those older versions. I'd like to ask whether there will be updates to improve the backward compatibility of responsive styles on older browsers because v4 is making my site completely unusable on older mobile browsers.

**Comment by mguzman-acgp** [comment]
I comment here to make it visible so more people find this discussion, I just found this asking to an LLM: it breaks the colors and gradients on iOS 15.6.1, even with tailwind v4.1.

**Comment by itsmikesharescode** [comment]
Here we go again.

**Reply by frederikhors** [reply]
?
